vendor/bat-native-ledger/src/bat/ledger/internal/wallet/wallet_recover.cc

```diff
@@ -0,0 +1,155 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "bat/ledger/internal/wallet/wallet_recover.h"
+
+#include <utility>
+
+#include "base/json/json_reader.h"
+#include "base/strings/string_split.h"
+#include "bat/ledger/internal/common/security_helper.h"
+#include "bat/ledger/internal/ledger_impl.h"
+#include "bat/ledger/internal/request/request_promotion.h"
+#include "bat/ledger/internal/request/request_util.h"
+#include "bat/ledger/internal/state/state_keys.h"
+#include "bat/ledger/internal/state/state_util.h"
+#include "net/http/http_status_code.h"
+
+#include "wally_bip39.h"  // NOLINT
+
+using std::placeholders::_1;
+
+namespace {
+
+ledger::Result ParseResponse(
+    const std::string& response,
+    std::string* payment_id) {
+  DCHECK(payment_id);
+
+  base::Optional<base::Value> value = base::JSONReader::Read(response);
+  if (!value || !value->is_dict()) {
+    return ledger::Result::LEDGER_ERROR;
+  }
+
+  base::DictionaryValue* dictionary = nullptr;
+  if (!value->GetAsDictionary(&dictionary)) {
+    return ledger::Result::LEDGER_ERROR;
+  }
+
+  const auto* payment_id_string = dictionary->FindStringKey("paymentId");
+  if (!payment_id_string || payment_id_string->empty()) {
+    BLOG(0, "Payment id is missing");
+    return ledger::Result::LEDGER_ERROR;
+  }
+
+  *payment_id = *payment_id_string;
+  return ledger::Result::LEDGER_OK;
+}
+
+}  // namespace
+
+namespace braveledger_wallet {
+
+WalletRecover::WalletRecover(bat_ledger::LedgerImpl* ledger) : ledger_(ledger) {
+}
+
+WalletRecover::~WalletRecover() = default;
+
+void WalletRecover::Start(
+    const std::string& pass_phrase,
+    ledger::ResultCallback callback) {
+  if (pass_phrase.empty()) {
+    BLOG(0, "Pass phrase is empty");
+    callback(ledger::Result::LEDGER_ERROR);
+    return;
+  }
+
+  auto phrase_split = base::SplitString(
+      pass_phrase,
+      " ",
+      base::TRIM_WHITESPACE,
+      base::SPLIT_WANT_NONEMPTY);
+
+  if (phrase_split.size() == 16) {
+    BLOG(0, "Pass phrase is niceware, update to bip39");
+    callback(ledger::Result::CORRUPTED_DATA);
+    return;
+  }
+
+  std::vector<unsigned char> new_seed;
+  new_seed.resize(32);
+  size_t written = 0;
+  const int result = bip39_mnemonic_to_bytes(
+      nullptr,
+      pass_phrase.c_str(),
+      &new_seed.front(),
+      new_seed.size(),
+      &written);
+
+  if (result != 0 || written == 0) {
+    BLOG(0, "Result: " << result << " Size: " << written);
+    callback(ledger::Result::LEDGER_ERROR);
+    return;
+  }
+
+  std::vector<uint8_t> secret_key =
+      braveledger_helper::Security::GetHKDF(new_seed);
+  std::vector<uint8_t> public_key;
+  std::vector<uint8_t> new_secret_key;
+  braveledger_helper::Security::GetPublicKeyFromSeed(
+      secret_key,
+      &public_key,
+      &new_secret_key);
+
+  const std::string public_key_hex =
+      braveledger_helper::Security::Uint8ToHex(public_key);
```

Comment:
Sorry only noticing this now, but it seems like we're doing this here and in `wallet_create.cc`. Is it worth pulling this out into a wallet utility function?

Reply:
good one, done